pkg/common-controller/snapshot_controller.go

```diff
@@ -533,19 +584,36 @@ func (ctrl *csiSnapshotCommonController) getPreprovisionedContentFromStore(snaps
 // A content is considered to be a pre-provisioned one if its Spec.Source.SnapshotHandle
 // is not nil, or a dynamically provisioned one if its Spec.Source.VolumeHandle is not nil.
 func (ctrl *csiSnapshotCommonController) getDynamicallyProvisionedContentFromStore(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotContent, error) {
+	driverName, err := ctrl.getSnapshotDriverName(snapshot)
+	if err != nil {
+		klog.Errorf("failed to getSnapshotDriverName while recording metrics for snapshot %q: %s", utils.SnapshotKey(snapshot), err)
+	}
+	createAndReadyOperation := metrics.NewOperation(metrics.CreateSnapshotAndReadyOperationName, driverName, snapshot)
+	snapshotStatusType := metrics.SnapshotStatusTypeControllerError
+	defer func() {
+		if err != nil {
+			// only emit if there is an error.
+			// success for createAndReadyOperation is emitted elsewhere.
+			ctrl.metricsManager.RecordMetrics(createAndReadyOperation, metrics.NewSnapshotOperationStatus(snapshotStatusType), snapshot)
+		}
+	}()
+
 	contentName := utils.GetDynamicSnapshotContentNameForSnapshot(snapshot)
 	content, err := ctrl.getContentFromStore(contentName)
 	if err != nil {
+		snapshotStatusType = metrics.SnapshotStatusTypeInvalidRequest
 		return nil, err
 	}
 	if content == nil {
 		// no matching content with the desired name has been found in cache
+		ctrl.metricsManager.RecordMetrics(createAndReadyOperation, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeInvalidRequest), snapshot)
```

Comment:
If you have `defer` above, you only need to set the snapshotStatusType?

Reply:
The defer only handles cases where err != nil. Otherwise, the defer would create ControllerError metrics on every reconcile without an error (even though the operation might not be finished)

Reply:
Got it.